### PR TITLE
GEP 1709: initial implementation ideas

### PR DESCRIPTION
**Pull request by shaneutt**
**What type of PR is this?**

/kind gep

**What this PR does / why we need it**:

This contributes to #1709 and #1329 by adding the initial ideas for an implementation of conformance profiles to the relevant GEP.

Please note that the GEP remains in `provisional` state for this iteration so nothing that merges here is necessarily set in stone. Therefore if you have some lingering questions or concerns there will be more time to voice those concerns, and ideally even make your own PR making the changes that you want to see and helping co-author the GEP to pull in your ideas!

### CODE REVIEW

File: site-src/geps/gep-1709.md
Comment (howardjohn) [inline site-src/geps/gep-1709.md:176]: can I be L7 and L4 at the same time?

Comment (howardjohn) [reply site-src/geps/gep-1709.md:176]: And mesh eventually

Comment (shaneutt) [reply site-src/geps/gep-1709.md:176]: There's a note about this concept here: https://github.com/kubernetes-sigs/gateway-api/blob/c6b5417042119e6d3e1b023c9ffd92935706f2f3/site-src/geps/gep-1709.md?plain=1#L103

I'll consider this one resolved, but if you feel like this is insufficient and you want to see more change I would definitely invite you to mark it unresolved and let me know what changes you feel are needed.

Comment (howardjohn) [reply site-src/geps/gep-1709.md:176]: I am not worried about the reporting but about the test running. Unless the tests for each profile are completely isolated so running all of them together vs running each profile in sequence is the ~same?

File: site-src/geps/gep-1709.md
Comment (LiorLieberman) [inline site-src/geps/gep-1709.md:56]: what is the existing command line approach? 
`go test  ......`?

File: site-src/geps/gep-1709.md
Comment (LiorLieberman) [inline site-src/geps/gep-1709.md:177]: Feels like we can combine `UnsupportedFeatures` and `SkipTests` to one field and have additional logic that checks whether this combined field is used with `Core` profile to check validity.

File: site-src/geps/gep-1709.md
Comment (LiorLieberman) [inline site-src/geps/gep-1709.md:315]: maybe a different colour for semi implemented Extended support?

File: site-src/geps/gep-1709.md
Comment (LiorLieberman) [inline site-src/geps/gep-1709.md:227]: One thing I have not found in the doc is who is responsible for generating `ConformanceReport` from the log?

Comment (LiorLieberman) [reply site-src/geps/gep-1709.md:227]: gw-api go library? 
a shell script?

I am thinking it will probably be the first one but probably worth calling it out

File: site-src/geps/gep-1709.md
Comment (LiorLieberman) [inline site-src/geps/gep-1709.md:293]: for this iteration, should we consider any kind of visualisation of conformance reports and/or badges other than looking into `conformance/results/<organization>-<project>.yaml` files?

Comment (LiorLieberman) [reply site-src/geps/gep-1709.md:293]: I think its better to keep it for future iterations but maybe worth adding a reference here

File: site-src/geps/gep-1709.md
Comment (LiorLieberman) [inline site-src/geps/gep-1709.md:75]: There are some impls that support experimental features, like istio and envoy-gw. 
Whats the additional cost of providing the profile for them to be able to include these tests?

File: site-src/geps/gep-1709.md
Comment (shaneutt) [reply site-src/geps/gep-1709.md:56]: Yes that's correct, we can add that here:

```suggestion
feasible and support both the existing command line integration approach (e.g. `go test`) as well
```

File: site-src/geps/gep-1709.md
Comment (shaneutt) [reply site-src/geps/gep-1709.md:315]: I'm open to doing something like this in the future if implementers complain about it... but I feel like extended is a completely reasonable place to have some features but not all.

File: site-src/geps/gep-1709.md
Comment (shaneutt) [reply site-src/geps/gep-1709.md:227]: I was actually thinking for the first iteration of this, these could just be manually edited as they have straightforward contents and need to go through a review process anyway. Let's make that more explicit:

```suggestion
For the above output upstream Gateway API can manually generate or update a report
in the following format:
```

File: site-src/geps/gep-1709.md
Comment (shaneutt) [reply site-src/geps/gep-1709.md:177]: I think `UnsupportedFeatures` and `SkipTests` need to be kept separate as `SkipTests` is particularly helpful _during_ the development process. E.g. someone could have partial support for a _feature_ but not all tests passing just yet and use `SkipTests` to disable the tests they're unable to fulfill just yet while they otherwise plan the work to fix it. We did literally this downstream at Kong and it was pretty convenient.

I'll consider this one resolved, but if you still feel there's more to do here please don't hesitate to unresolve and let me know if I've missed something or there's something else you feel needs changing.

File: site-src/geps/gep-1709.md
Comment (shaneutt) [reply site-src/geps/gep-1709.md:294]: I think that would be good, however I'm not sure _what_ that should look like yet. I'm actually hoping that someone else will swoop in and think up a display layer using the raw data laid out so far.

File: site-src/geps/gep-1709.md
Comment (arkodg) [inline site-src/geps/gep-1709.md:156]: why not add a section in the report highlighting what was `Skipped` ?

File: site-src/geps/gep-1709.md
Comment (shaneutt) [reply site-src/geps/gep-1709.md:75]: In my first pass thinking through this I honestly didn't really think experimental features needed any special recognition, as they have a special nature that makes them off the normal course of the project's feature scope. That said, if there's some strong feelings we need to represent _experimental feature support_ visibly at a high level with conformance profiles I am open to that, I'm just not quite convinced yet :thinking:

File: site-src/geps/gep-1709.md
Comment (arkodg) [reply site-src/geps/gep-1709.md:176]: what about adding the ability to input a list of supported Profiles ?

File: site-src/geps/gep-1709.md
Comment (shaneutt) [reply site-src/geps/gep-1709.md:156]: My intention here was to _never_ report conformance when tests are explicitly skipped, instead making `SkipTests` a tool for the development phase. Can you help me to understand the value you see in allowing reporting where entire tests are skipped, even if they might have been required? In the case of extended where we know some _features_ may not be fully supported, wouldn't `UnsupportedFeatures` adequately cover things here?

Comment (arkodg) [reply site-src/geps/gep-1709.md]: `SkipTests` are fewer, more fine grained tests, within `SupportedFeatures`, that the project cannot support either due to a bug in the project or due to API clarification. Here the setting in [Envoy Gateway](https://github.com/envoyproxy/gateway/blob/afe646b84d03cf4126f55721589850dc202dc130/test/conformance/conformance_test.go#L50). 

With above decision, Envoy Gateway would continue to run all tests, except some with SkipTests (temporarily until issues are fixed), but would not be able to generate any report

File: site-src/geps/gep-1709.md
Comment (arkodg) [reply site-src/geps/gep-1709.md]: looks good, thanks !

File: site-src/geps/gep-1709.md
Comment (shaneutt) [reply site-src/geps/gep-1709.md]: Right, I think we're kinda saying the same thing. But where I was kinda coming from, is that if you're _skipping tests_ then this basically means you're **not** conformant. I'm a little confused about having a reporting option for implementations which are not yet conformant, but I guess I could be persuaded? :thinking:

Comment (arkodg) [reply site-src/geps/gep-1709.md]: Sure from a reporting standpoint, you can mark it as `not` conformant :)
From an implementation perspective, we prefer this approach to ensure more test coverage and minimal regression
TLDR, would prefer if the reporting tool marks the specific features (whose tests are skipped) as non conformant rather than fail reporting for other features too

File: geps/gep-1709.md
Comment (sunjayBhatia) [inline geps/gep-1709.md:231]: worth versioning this schema? in case of future changes that might need a migration etc.?

File: site-src/geps/gep-1709.md
Comment (arkodg) [reply site-src/geps/gep-1709.md]: The conformance tests run in CI in EG to prevent regression as well as report conformance (e.g. X feature tests pass). Earlier I was thinking we could this plug into the CI and run the tests once , so the report is generated on every PR run, so EG would replace this suite flag `EnableAllSupportedFeatures: true,` with `Profile: Layer4CoreProfile,Layer4ExtendedProfile,Layer7CoreProfile,Layer7ExtendedProfile`. 

But thinking more on this, the conformance profile tests might need to run as a separate task who's job is to only report conformance rather than use the exit code of the process to decide whether CI is passing or not (catch regression).

File: geps/gep-1709.md
Comment (shaneutt) [reply geps/gep-1709.md:226]: Yes, I was just thinking about this today too!

Added in b783fcac4db87482cc0f9f43e19828e237d01de3 :+1:

File: site-src/geps/gep-1709.md
Comment (shaneutt) [reply site-src/geps/gep-1709.md]: I added a TODO comment to the document that says we must figure this out _BEFORE_ we consider this `implementable`. That way we don't have to hold up this iteration just for that bit. :+1:

File: site-src/geps/gep-1709.md
Comment (shaneutt) [reply site-src/geps/gep-1709.md]: After talking with @LiorLieberman in a video call, the entire concept of how you refer to profiles has been reworked to incorporate providing a positive (or negative) list of additional features so this should be set now.

File: site-src/geps/gep-1709.md
Comment (shaneutt) [reply site-src/geps/gep-1709.md]: So it just occurred to me a couple days ago, I had already added a note about being able to explicitly report "in progress" conformance via this process:

https://github.com/kubernetes-sigs/gateway-api/pull/1754/files#diff-7764ccb47f68a80348b14a2c6df6c6cadd80db02de7817dbcb8039a611bf1328R130

Also I've removed the support levels on the profiles in favor of a nested reporting structure. If I'm understanding what you were asking for, I think this will provide the facility you were looking for. If not, please mark this unresolved and let me know!

File: geps/gep-1709.md
Comment (LiorLieberman) [inline geps/gep-1709.md:95]: typo?

File: geps/gep-1709.md
Comment (LiorLieberman) [inline geps/gep-1709.md:97]: `[Golang library][lib]` typo?

File: geps/gep-1709.md
Comment (sunjayBhatia) [inline geps/gep-1709.md:288]: might want to add somewhere that we should have CI in this repo to do some YAML linting and/or schema validation of these uploaded reports

implementation detail but maybe good to point out as a requirement if we're starting to do this with "hand crafted" YAML

File: geps/gep-1709.md
Comment (sunjayBhatia) [inline geps/gep-1709.md:230]: wdyt about each project/implementation's report file being a list of these sort of report objects? e.g.

```yaml
---
kind: ConformanceReport
apiVersion: v1alpha1
implementation:
  name: acmeorg-acme
  url: https://acmeorg.com
  version: v9.99.0
  contact:
  - @acme-org/maintainers
date: "2023-02-28 20:29:41+00:00"
gatewayAPIVersion: v0.6.1
profiles:
- name: layer4
  core:
    supportedFeatures:
    - ...
  extended:
    supportedFeatures:
    - ...
    unsupportedFeatures:
    - ...
- name: layer7
  core:
    supportedFeatures:
    - ...
  extended:
    supportedFeatures:
    - ...
    unsupportedFeatures:
    - ...
---
kind: ConformanceReport
apiVersion: v1alpha1
implementation:
  name: acmeorg-acme
  url: https://acmeorg.com
  version: v9.98.0
  contact:
  - @acme-org/maintainers
date: "2023-02-28 20:29:41+00:00"
gatewayAPIVersion: v0.6.0
profiles:
...
```

this would allow us to add the fields like date, version, etc. to each report as well since that detail is often relevant when assessing an implementation

File: geps/gep-1709.md
Comment (shaneutt) [inline geps/gep-1709.md:279]:
```suggestion
TODO: need to add testing flags to emit the `ConformanceReport` here automatically

The implementation can compile its `ConformanceReport` and upload it to the
```

File: geps/gep-1709.md
Comment (shaneutt) [reply geps/gep-1709.md:95]: I don't think so, this just provides a hyperlink which is handled further down. Please un-resolve and let me know if there's still a problem though? :thinking:

File: geps/gep-1709.md
Comment (shaneutt) [reply geps/gep-1709.md:97]: Similar to above, this just adds a hyperlink which appears to be working? Please un-resolve and let me know if there's still a problem though? :thinking:

File: geps/gep-1709.md
Comment (shaneutt) [reply geps/gep-1709.md:289]: sg, added a TODO for this in cfcc08351432dfe239ff18dd295ff357f55716ef, we will need to resolve this and other TODO's prior to going `implementable` :+1:

File: geps/gep-1709.md
Comment (shaneutt) [reply geps/gep-1709.md:216]: We may have been thinking this separately and simultaneously, because I just recently added a TODO for this:

75ec95f8e573dc6602020c3b8ecb99781bed75df

Let me know if that seems adequate :+1:

File: geps/gep-1709.md
Comment (shaneutt) [reply geps/gep-1709.md:229]: I can see advantages to this, particularly for the purpose of the human reporting via a PR part, and also making it really straightforward to build an archive several years down the road when these reports get significantly bigger. :thinking: 

Only thing I would say is that you've combined `supportedFeatures` and `unsupportedFeatures` in the same report for `extended`, but there is a note that had attempted to establish these as mutually exclusive earlier. We'll have to discuss more on the subject of how we represent extended features (and maybe even implementation-specific ones) but that doesn't need to hold this particular comment up.

@LiorLieberman and @arkodg had also expressed interest in contributing to this doc, so I would like to hear from there.

Comment (sunjayBhatia) [reply geps/gep-1709.md:229]: ah yeah feel free to disregard that part on the reporting of what is actually passed/failed etc. that was just a quick idea without considering what may have already been updated here

but good to hear some of that metadata additions and the idea that we could generate a report per-run of the suite sounds useful 👍🏽

File: geps/gep-1709.md
Comment (sunjayBhatia) [reply geps/gep-1709.md:216]: yep that sounds great 👍🏽

File: geps/gep-1709.md
Comment (sunjayBhatia) [reply geps/gep-1709.md:229]: I also considered suggesting something of the form of what sonobuoy or some other tool might output, but not sure if we want to go that far in reporting test result details etc., I don't remember exactly but I imagine using sonobuoy was discussed when starting the conformance framework

File: geps/gep-1709.md
Comment (mlavacca) [inline geps/gep-1709.md:302]: I think we could at least check that if an implementor sets a specific profile, all the tests belonging to such a profile are successfully executed. So, for example, if not running specific tests belonging to a profile through the `SkipTests` field, we might want to invalidate the certification as some required tests have not been run. WDYT?

File: geps/gep-1709.md
Comment (shaneutt) [reply geps/gep-1709.md:229]: I wouldn't be opposed to using that in time, if we found there were advantages. But at least for this very specific PR I'm not ready to completely switch things around yet. I would encourage you to make a PR after this one merges _just_ for pushing this change so that we can hyper-focus on it.

File: geps/gep-1709.md
Comment (shaneutt) [reply geps/gep-1709.md:302]: > I think we could at least check that if an implementor sets a specific profile, all the tests belonging to such a profile are successfully executed.

Yes, but more specifically all the _core_ tests. We are already making _core_ a minimum requirement:

https://github.com/kubernetes-sigs/gateway-api/pull/1754/files#diff-7764ccb47f68a80348b14a2c6df6c6cadd80db02de7817dbcb8039a611bf1328R169

> So, for example, if not running specific tests belonging to a profile through the SkipTests field, we might want to invalidate the certification as some required tests have not been run. WDYT?

Yes. There's a note specifically about this present here:

https://github.com/kubernetes-sigs/gateway-api/pull/1754/files#diff-7764ccb47f68a80348b14a2c6df6c6cadd80db02de7817dbcb8039a611bf1328R127

We're not 100% sure exactly how it will _look_ yet, but we are already trying to be clear that `SkipTests` usage will invalidate the `ConformanceReport` (or at least mark conformance as "in-progress") or something like that.

Comment (mlavacca) [reply geps/gep-1709.md:302]: Thanks for pointing out the notes mentioning this aspect, I somehow lost them :)

File: geps/gep-1709.md
Comment (arkodg) [reply geps/gep-1709.md:229]: my high level comment here is, we should separate out `what tests to run` with `what to report`.
* under `what tests to run` , we should continue to support the existing knobs of `SupportedFeatures`, `UnsuportedFeatures`, `SkipTests` and `SupportAllFeatures`. The user can either choose to use these historic knobs **OR** use the new `ConformanceProfile` knob. These flags should be validated.
* under `what to report`, it could be as granular as which tests passed or more coarse as which features passed (if all tests within that feature passed), but this logic is not tied to the input knobs, and is tied to - which tests ran, which passed, which failed. And the report is an explicit add on data output enabled via some `--report` like flag

Comment (shaneutt) [reply geps/gep-1709.md:229]: what you're suggesting seems to at least be close to where we're headed already @arkodg, do you feel there's anything specific in the doc that needs to be changed?

Comment (arkodg) [reply geps/gep-1709.md:229]: yeah the doc says `Profiles` flag, in conjunction with `SkipTests` will result in an invalid report. Prefer if this tied to `what tests to run` instead, and said either use `Profile`  or use a mix of `SupportedFeatures, UnsuportedFeatures, SkipTests and SupportAllFeatures` to decide which tests to run. If not adhered, the tests wont start/or fail....

Comment (arkodg) [reply geps/gep-1709.md:229]: ^ can be added later as well as a follow up

Comment (sunjayBhatia) [reply geps/gep-1709.md:229]: regardless of the exact reporting format of test results, seems useful to generally summarize and agree on somewhere in this doc what is useful for gw api maintainers/implementers/users to see in the conformance test report

my two cents:
- the project/version/test run metadata above to identify a particular test result ^
- conformance profile test run results overall summary for each "profile"
  - implementation is deemed fully conformant with the profile if the tests all passed 
  - partial if some passed/some were skipped (IMO this is useful to know for esp. the extended profile)
  - no conformance if none passed or none were run (i.e. an L4 only controller might skip all L7 tests)
  - include summary stats, e.g. total number of tests, how many passed, skipped, failed
- conformance profile test run detailed results (in case of partial conformance etc.)
  - for each test, test name, short description like we have [here](https://github.com/kubernetes-sigs/gateway-api/blob/69fabd994640debc2d9c00d4a086cdee0bbe9a22/conformance/tests/gateway-invalid-route-kind.go#L35-L36)
  - did it pass, fail, or was it skipped
  - can include test filename for people to do closer inspection if needed
  - can add detailed failure messages, etc. if we want too

Comment (sunjayBhatia) [reply geps/gep-1709.md:229]: (IMO seems ok to skip tests, that shouldn't necessarily make the output "invalid" but will mean you end up with a report that your implementation is not fully conformant)

Comment (sunjayBhatia) [reply geps/gep-1709.md:229]: I'm on board with the notion of separating what to test from what to report conceptually as @arkodg mentions above, we could e.g. start with no changes to the test suite invocation and iteratively add convenience flags to let you run subsets of tests that correspond to "profiles"

**Comment by shaneutt** [comment]
Marking this to not automatically merge, I will merge this manually once it's got the final LGTM.

File: geps/gep-1709.md
Comment (shaneutt) [reply geps/gep-1709.md:223]: After thinking it over I was persuaded by the list of separate `ConformanceReports` idea, and did a reworking of the `ConformanceReport` API which includes that and somewhat follows the bullet points you listed above, @sunjayBhatia.

My hope is that this will be satisfactory for all stakeholders. I'm going to consider this comment thread resolved, as it's gotten quite long and hard to follow, but if any of you _don't_ find the new version satisfactory please create some new comments **ideally with explicitly and prescriptive suggestions about what specifically you would like to see changed** as that can be incredibly helpful to keeping up velocity on these GEPs. Also keep in mind that we're still in a `provisional` state so nothing that merges here is going to be implemented yet.

**Review by arkodg — APPROVED** [review]
this is a good start, prefer to get this in, and iteratively improve this doc, thanks !

**Comment by shaneutt** [comment]
@sunjayBhatia I'm marking this one as a merge hold for the moment to make sure we hear back from you as a stakeholder before we merge it.

**Comment by LiorLieberman** [comment]
I like the changes. Thanks!

/lgtm
/approve

**Review by sunjayBhatia — APPROVED** [review]
looks great to me 👍🏽 

I think regardless of how the implementation details etc. shake out (in terms of reporting schema, command line flags, etc.) I think the core ideas of what we want to get out of the centralized profiles/reporting are there!

**Comment by shaneutt** [comment]
@youngnick as I understand it you wanted to give this a pass too, so keeping the hold on it until you've had a chance. Please feel free to remove the hold now once you're happy with it :+1:

**Review by youngnick — COMMENTED** [review]
This LGTM, I think a couple of things I consider imperative are done:
- Using `SkipTests` on any Core features means you're not conformant
- The output is machine parsable.

The one thing I'd add, non-blocking though, is an example of what the YAML would look like with failed tests. Not that I think anyone would submit that, but it will make it clearer for implementors.

**Comment by shaneutt** [comment]
> * Using `SkipTests` on any Core features means you're not conformant
> * The output is machine parsable.
> * The one thing I'd add, non-blocking though, is an example of what the YAML would look like with failed tests.

These things are already accounted for in the document:

https://github.com/kubernetes-sigs/gateway-api/pull/1754/files#diff-7764ccb47f68a80348b14a2c6df6c6cadd80db02de7817dbcb8039a611bf1328R127
https://github.com/kubernetes-sigs/gateway-api/pull/1754/files#diff-7764ccb47f68a80348b14a2c6df6c6cadd80db02de7817dbcb8039a611bf1328R369

> Oh, and it would be good to add that the descriptions are free-form, we shouldn't require specific values there.

This one wasn't accounted for yet, I've added this in 6489c36356345789354402103cfe9976a466690a :+1:

**Pull request merged**
